File: find peek.java
// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

class Solution {
    public int findPeakElement(int[] nums) {
        int l = 0;
        int h = nums.length - 1;
        while (l<=h){
            int mid = l + (h-l)/2;
            if(nums == null &&  nums.length == 0) return -1;
            if(nums.length == 1) return 0;
            if(0 == mid && (nums[mid] > nums[1])) return mid;
            if((nums.length - 1 == mid) && (nums[mid] > nums[mid - 1])) return mid;
            if ((nums[mid] > nums[mid + 1]) && (nums[mid] > nums[mid - 1])) return mid;
            if(nums[mid] < nums[mid + 1]){
                l = mid + 1;
            }
            else{
                h = mid - 1;
            }
        }
    return -1;
    }
}

File: findPeek.py
#Time Complexity: O(log n)
#Space Complexity : O(1) as we only use pointers high, low,mid
#did your solution run on leetcode - yes

class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        low = 0
        high = len(nums) - 1
        
        while (low <= high):
            mid = low + (high - low)//2
            #checking edge cases and also checking if the mid is greater than both of its neighbours if yes mid is peak
            if ((mid == 0 or (nums[mid - 1] < nums[mid])) and (mid == len(nums)-1 or nums[mid]> nums[mid + 1]  )):
                return mid
            
            #check whether next > mid move towards the right half of list
            elif  nums[mid] < nums[mid + 1]:
                low = mid + 1
            
            #check whether prev > mid move towards the left half of list
            elif  nums[mid - 1] > nums[mid]:
                high = mid - 1
        return 0
       
        
